### ISSUE TITLE
Feature request for `[:<>]`

### ISSUE THREAD

**Issue by agzam** [weavejester/hiccup#164]
Would it be possible to add a feature similar to Re-agent's support for React Fragments. So it is possible to do things like this:

    (def content
      [:<>
        [:li "one"]
        [:li "two"]
        [:li "three"]])

    (html5 
      [:ul content])
      
Currently you have to do something like:

    (def content
      [[:li "one"] [:li "two"] [:li "three"]])

    (html5 
      [:ul (map identity content)])
      
Which is a bit difficult, now the caller has to know about the structure of sub-component

**Comment by weavejester** [comment]
You can just return a seq directly:

```clojure
(def content
  (list
   [:li "one"]
   [:li "two"]
   [:li "three"]))

**Comment by agzam** [comment]
Thank you James. Somehow trying that didn't occur to me.

**Comment by mhuebert** [comment]
When targeting react w/ reagent, something like `[:<> child1 child2]` and `(list child1 child2)` represent the same DOM elements but they are treated differently by React - the former doesn't require a unique key for each child, whereas the latter does. Supporting `:<>` on the server (in this lib) would make it easier to write code that works in both environments.

**Comment by shaunlebron** [comment]
[rum](https://github.com/tonsky/rum) can be used to react-specific hiccups (with `:<>` fragments) on server side via `rum.core/render-html`.

**Comment by timothypratley** [comment]
Hi, I'd like to request this feature and suggest reopening the issue.
My use case is that [nextjournal markdown](https://github.com/nextjournal/markdown) produces fragments `[:<> ...]`.
As a workaround I traverse the hiccup and replace them with seqs, so it's not urgent but fwiw I do think this would be a convenient and easy feature when dealing with fragments.

**Comment by weavejester** [comment]
> My use case is that [nextjournal markdown](https://github.com/nextjournal/markdown) produces fragments `[:<> ...]`.

Wouldn't a better solution be to fix the output from nextjournal markdown instead?

**Comment by timothypratley** [comment]
IMO supporting fragments in hiccup is better.

Nextjournal should produce `[:<> ...]` because they target ClojureScript/Reagent style applications.
Seqs and fragments have slightly different meanings (seqs require ids) in such contexts.
Producing seqs is not a good default.

On the flip side, I don't see a downside to supporting fragments in hiccup.

**Comment by timothypratley** [comment]
Another data point: https://github.com/lambdaisland/hiccup
lists as it's #2 feature fragment support - perhaps an indication that it was a desirable feature

**Comment by weavejester** [comment]
> Nextjournal should produce `[:<> ...]` because they target ClojureScript/Reagent style applications.

Then shouldn't that be an option? I feel that compatibility should be the default, or at least *possible* via an option.

I'm generally cautious of supporting additional syntax, as we cannot remove support for it later. I'm also reluctant to support two different syntaxes that have the same semantic meaning.

**Comment by timothypratley** [comment]
Yes, I agree with and understand the need for caution.
This library is the most widely used and loved Clojure library (congratulations! that's awesome)
And yes, I have many alternatives for solving my particular use case(s).

The purpose of me asking you to reconsider this feature is that I believe it will be a small but welcome improvement to your library and benefit many people.

Hiccup usage has become prolific on the frontend. For myself and others the syntax of hiccup has already moved slightly beyond the definition provided by this (excellent) library as evidenced by Reagent/RUM/etc, several competing implementations of hiccup that generally attempt to support cross-platform CLJS/CLJ, functions and fragments as features to enable more use cases such as server-side rendering, interactive notebooks and markdown parsing. These use cases exist and are satisfied by other libraries. The differences are very small. The "expanded definition" of frontend compatible hiccup has not evolved much in the last 5 years;  "functions as tags", "fragments", and "className for class" sums it up.

2 syntaxes already exist in expanded hiccup for sequences and fragments. Specifically the difference is "a sequence of elements that have keys to enable updates" (sequence) vs "children without keys" (fragment). This difference is important for reactive updates and is the reason that the new spaceship tag syntax was introduced.
For static HTML it is not necessary to check for or enforce keys be present.
The value in supporting the spaceship is in being compatible.

I don't see a downside to increasing compatibility.
In this particular instance, `[:<> ...]` currently produces invalid HTML.
Making it valid reserves the spaceship tag as a first class member of hiccup syntax, making it unavailable for alternative purposes. It's hard to imagine an alternative that makes sense, so it seems like a small risk.

I ask that you consider that this is indeed a feature that at least a few people have asked for and weigh the cost/value of it. If you've already decided not to include it, I'll happily accept that answer.

**Comment by weavejester** [comment]
I think before I can even make a decision, I need to understand why there are two syntaxes in Reagent, as I'm not clear on what the purpose and use cases are. One requires keys and the other doesn't, but why? Clearly at some point it was decided that it would be useful in Reagent, but in React there appears to be only one syntax for fragments.

**Comment by timothypratley** [comment]
They are both React concepts, and the only difference is performance.

Fragments `[:<> ...]` are described [here](https://react.dev/reference/react/Fragment)

Sequences (arrays in JavaScript) are described [here](https://react.dev/learn/rendering-lists#rendering-data-from-arrays)

[why-does-react-need-keys](https://react.dev/learn/rendering-lists#why-does-react-need-keys) does a good job of explaining the performance implications.

TLDR: updating by index is inefficient (consider a list where an item moves position within the list).

FWIW it's a subtle and somewhat silly distinction (let's make an efficient way that nobody will use, and an inefficient way that everyone will love), *shrug* but that's what they did. I could be missing the point, but that's the background AFAIK.

**Comment by weavejester** [comment]
Ah, that makes more sense to me. So Reagent is just mapping the concepts of React onto Hiccup structures.

My inclination is still not to mix React-isms with Hiccup but I'll give it some thought.

Another factor is that there are enough issues with Hiccup 2.0's design that I'm considering a 3.0 version (or a 'next' version), where rendering and compiling are disentangled somehow. Potentially this could take options that would allow for Reagent's extensions.

**Comment by timothypratley** [comment]
> It's hard to imagine an alternative

I've imagined a scenario that seems obvious in hindsight... maybe someone would want to produce `<></>` if they were going to subsequently feed this to React in some way 🤔
